### PR TITLE
use context manager for path extension in torch.hub

### PR DESCRIPTION
**Pull request by pmeier**
We are using the idiom

```py
sys.path.insert(0, path)

# do something

sys.path.remove(path)
```

three times in `torch.hub`. This is a textbook case for using a context manager. In addition, by using `try` / `finally` we can enforce the Python path is back in its original state even if the actual action raises an exception:

```py
import sys


path = "/tmp"

# PR
try:
    sys.path.insert(0, path)
    try:
        # Any exception raised while performing the actual functionality
        raise Exception
    finally:
        sys.path.remove(path)
except Exception:
    assert path not in sys.path

# main
try:
    sys.path.insert(0, path)

    # Any exception raised while performing the actual functionality
    raise Exception

    sys.path.remove(path)
except Exception:
    assert path in sys.path
```

cc @nairbv @NicolasHug @vmoens @jdsgomes

### CODE REVIEW

File: torch/hub.py [hunk 1]
```diff
@@ -545,15 +549,12 @@ def _load_local(hubconf_dir, model, *args, **kwargs):
...
-
-    entry = _load_entry_from_hubconf(hub_module, model)
-    model = entry(*args, **kwargs)
+    with _add_to_sys_path(hubconf_dir):
+        hubconf_path = os.path.join(hubconf_dir, MODULE_HUBCONF)
+        hub_module = _import_module(MODULE_HUBCONF, hubconf_path)
 
-    sys.path.remove(hubconf_dir)
+        entry = _load_entry_from_hubconf(hub_module, model)
+        model = entry(*args, **kwargs)
```
Comment (pmeier) [inline torch/hub.py:557]: We can probably move this out of the context manager, given that `torch.hub.help` also uses `_load_entry_from_hubconf` after the path is already reset (L446).

**Comment by NicolasHug** [comment]
Thanks Philip, changes look good. Let's merge time after we get a chance to check the changes from https://github.com/pytorch/pytorch/pull/72060. I tried to do that a few times, but the torchhub repo's CI is toasted currently...

**Comment by NicolasHug** [comment]
I promise I'll get to it one day <3

**Review by NicolasHug — APPROVED** [review]
Thanks a lot Philip for the PR and for your patience!
Changes LGTM, now that's the torchhub repo CI works properly again we can safely merge this.

**Comment by pmeier** [comment]
Test failure is unrelated: https://github.com/pytorch/pytorch/actions/runs/3862444806/jobs/6584292375#step:10:44673

@pytorchbot merge -f "unrelated trunk failure"

**Pull request closed without merging**